### PR TITLE
Replace ignition templates with go reconciliation code

### PR DESCRIPTION
**Pull request by csrwng**
Replaces template-based manifests that generate ignition config with go
code using typed structs. Scripts and systemd unit file remain as templates.

Vendors alternate json serializer used by the installer to serialize ignition configs.
Vendors data structure for ignition config only from coreos/ignition repo. 

Bumps the ignition version in use to 3.2, since that is what the installer uses in 4.9.

### CODE REVIEW

File: thirdparty/ignition/v3_2/schema.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
+package types
```
Comment (enxebre) [inline thirdparty/ignition/v3_2/schema.go:1]: ignition is vendored here https://github.com/openshift/hypershift/blob/31ec41b3aa2a670c2a8dbfbe93634de97ad1ff85/hypershift-operator/controllers/nodepool/nodepool_controller.go#L623

https://github.com/openshift/hypershift/tree/main/vendor/github.com/coreos/ignition/v2

Comment (csrwng) [reply thirdparty/ignition/v3_2/schema.go:1]: thx! will remove the extra one

Comment (csrwng) [reply thirdparty/ignition/v3_2/schema.go]: updated to use existing vendored package, but switch to v3.2 which is what the installer uses in 4.9+

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 2]
```diff
@@ -706,6 +707,12 @@ func (r *HostedControlPlaneReconciler) update(ctx context.Context, hostedControl
 		return fmt.Errorf("failed to reconcile olm: %w", err)
 	}
 
+	// Reconcile Ignition
+	r.Log.Info("Reconciling Ignition")
+	if err = r.reconcileIgnition(ctx, hostedControlPlane, releaseImage, infraStatus.APIHost, infraStatus.APIPort); err != nil {
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:712]: may be call this reconcileCoreIgnitionConfig / reconcileCoreMachineConfig ?

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 3]
```diff
@@ -849,7 +849,7 @@ func getAMI(nodePool *hyperv1.NodePool, region string, releaseImage *releaseinfo
 func ignConfig(encodedCACert, encodedToken, endpoint string) ignitionapi.Config {
 	return ignitionapi.Config{
 		Ignition: ignitionapi.Ignition{
-			Version: "3.1.0",
+			Version: "3.2.0",
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:852]: curious any particular motivation for the bump or just being at latest?

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 4]
```diff
@@ -2198,23 +2205,57 @@ func (r *HostedControlPlaneReconciler) reconcileOperatorLifecycleManager(ctx con
...
 		data, hasSSHKeyData := sshKeySecret.Data["id_rsa.pub"]
 		if !hasSSHKeyData {
-			return nil, fmt.Errorf("SSH key secret secret %s is missing the id_rsa.pub key", hcp.Spec.SSHKey.Name)
+			return fmt.Errorf("SSH key secret secret %s is missing the id_rsa.pub key", hcp.Spec.SSHKey.Name)
 		}
-		sshKeyData = data
+		sshKey = string(data)
+	}
+
+	p := ignition.NewIgnitionParams(hcp, releaseImage.ComponentImages(), apiServerAddress, apiServerPort, sshKey)
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:2224]: same than https://github.com/openshift/hypershift/pull/668/commits/8b0b3e7a606ce7fbd87cb811ed8d12aa2b5dfd3a#r748203820

I'd call this IgnitionConfig consistently everywhere.

File: control-plane-operator/controllers/hostedcontrolplane/ignition/reconcile.go [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+	jsonBytes, err := json.Marshal(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling ignition config: %w", err)
+	}
+	return jsonBytes, nil
+}
+
+func setMachineConfigLabels(mc *mcfgv1.MachineConfig) {
+	mc.Labels = map[string]string{
+		"machineconfiguration.openshift.io/role": "worker",
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/ignition/reconcile.go:254]: should we put this into a gloabal var?

File: control-plane-operator/controllers/hostedcontrolplane/ignition/reconcile.go [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+	buf := &bytes.Buffer{}
+	mc.APIVersion = mcfgv1.SchemeGroupVersion.String()
+	mc.Kind = "MachineConfig"
+	if err := api.YamlSerializer.Encode(mc, buf); err != nil {
+		return fmt.Errorf("failed to serialize machine config %s: %w", cm.Name, err)
+	}
+	if cm.Labels == nil {
+		cm.Labels = map[string]string{}
+	}
+	cm.Labels["hypershift.openshift.io/core-ignition-config"] = "true"
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/ignition/reconcile.go:269]: should we put this into a gloabal var?

**Comment by enxebre** [comment]
Thanks @csrwng! dropped some minor comments.
/lgtm
/hold

feel free to cancel the hold.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 3 above)
Comment (csrwng) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:852]: In 4.9 and above we are using 3.2.0 -> https://github.com/openshift/installer/blob/1c538b8949f3a0e5b993e1ae33b9cd799806fa93/pkg/asset/ignition/bootstrap/common.go#L21

**Comment by csrwng** [comment]
Thanks @enxebre. If ok with you I can address your comments in a follow-up PR. I would like to merge this one and rebase the other ones on top of it.

**Pull request merged**